pkg/udf/map_udf.go

```diff
@@ -45,149 +46,176 @@ func (u *MapUDFProcessor) Start(ctx context.Context) error {
+	c, err := clientsdk.New(clientsdk.WithMaxMessageSize(maxMessageSize))
+	if err != nil {
+		return fmt.Errorf("failed to create a new gRPC client: %w", err)
+	}
+	udfHandler, err := function.NewUDSgRPCBasedUDF(c)
+	if err != nil {
+		return fmt.Errorf("failed to create gRPC client, %w", err)
+	}
+	// Readiness check
+	if err := udfHandler.WaitUntilReady(ctx); err != nil {
+		return fmt.Errorf("failed on UDF readiness check, %w", err)
+	}
+	defer func() {
+		err = udfHandler.CloseConn(ctx)
+		if err != nil {
+			log.Warnw("Failed to close gRPC client conn", zap.Error(err))
+		}
+	}()
+
+	// create readers and writers
+	var readers []isb.BufferReader
+	var writers map[string][]isb.BufferWriter
 	// watermark variables
 	fetchWatermark, publishWatermark := generic.BuildNoOpWatermarkProgressorsFromBufferList(u.VertexInstance.Vertex.GetToBuffers())
 
 	switch u.ISBSvcType {
 	case dfv1.ISBSvcTypeRedis:
-		reader, writers = buildRedisBufferIO(ctx, fromBufferName, u.VertexInstance)
+		readers, writers, err = buildRedisBufferIO(ctx, u.VertexInstance)
+		if err != nil {
+			return err
+		}
 	case dfv1.ISBSvcTypeJetStream:
 		// build watermark progressors
+		// multiple go routines can share the same set of writers since nats conn is thread safe
+		// https://github.com/nats-io/nats.go/issues/241
 		fetchWatermark, publishWatermark, err = jetstream.BuildWatermarkProgressors(ctx, u.VertexInstance)
 		if err != nil {
 			return err
 		}
-		reader, writers, err = buildJetStreamBufferIO(ctx, fromBufferName, u.VertexInstance)
+		readers, writers, err = buildJetStreamBufferIO(ctx, u.VertexInstance)
 		if err != nil {
 			return err
 		}
 	default:
 		return fmt.Errorf("unrecognized isbsvc type %q", u.ISBSvcType)
 	}
 
-	// Populate shuffle function map
-	shuffleFuncMap := make(map[string]*shuffle.Shuffle)
-	for _, edge := range u.VertexInstance.Vertex.Spec.ToEdges {
-		if edge.ToVertexType == dfv1.VertexTypeReduceUDF && edge.GetToVertexPartitions() > 1 {
-			s := shuffle.NewShuffle(u.VertexInstance.Vertex.GetName(), edge.GetToVertexPartitions())
-			shuffleFuncMap[fmt.Sprintf("%s:%s", edge.From, edge.To)] = s
+	for index, bufferPartition := range fromBuffer {
+		// Populate shuffle function map
+		shuffleFuncMap := make(map[string]*shuffle.Shuffle)
+		for _, edge := range u.VertexInstance.Vertex.Spec.ToEdges {
+			if edge.ToVertexType == dfv1.VertexTypeReduceUDF && edge.GetToVertexPartitions() > 1 {
+				s := shuffle.NewShuffle(u.VertexInstance.Vertex.GetName(), edge.GetToVertexPartitions())
+				shuffleFuncMap[fmt.Sprintf("%s:%s", edge.From, edge.To)] = s
+			}
 		}
-	}
 
-	conditionalForwarder := forward.GoWhere(func(keys []string, tags []string) ([]forward.VertexBuffer, error) {
-		var result []forward.VertexBuffer
+		// create a conditional forwarder for each partition
+		getVertexPartition := GetPartitionedBufferName()
+		conditionalForwarder := forward.GoWhere(func(keys []string, tags []string) ([]forward.VertexBuffer, error) {
```

Comment:
could all partitions use the same conditionalForwarder? I was trying to see what was different for each, but I may have missed it.

Reply:
` GetPartitionedBufferName()` is separate for each partition

Reply:
It looks like `GetPartitionedBufferName()` is set to the same function each time? It seems to me that in order for there to be some difference in all of these ConditionalForwarders, I'd need to see `index` or `bufferPartition` (which we're iterating through) used somewhere on those lines. Now, on the other hand, if `conditionalForwarder` requires state (it's something that gets modified in the course of running), then it's reasonable to have separate instances of them. Anyway, it's not a big deal either way.

Reply:
never mind, I see the state is the map defined in GetPartitionedBufferName()